viewmodels/src/androidMain/kotlin/com/mirego/trikot/viewmodels/AndroidViewModelProviderFactory.kt

```diff
@@ -0,0 +1,198 @@
+package com.mirego.trikot.viewmodels
+
+import android.content.Intent
+import android.os.Bundle
+import androidx.fragment.app.Fragment
+import androidx.fragment.app.FragmentActivity
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.ViewModelProvider
+import kotlin.reflect.KClass
+
+/**
+ * Android extensions around a ViewModelProvider which uses reflection to instanciate
+ * view models and optionally pass it a string argument. The argument is passed to the activity
+ * intent and the fragment arguments.
+ *
+ * Terminology: viewModelController is simply the root view model for the activity / fragment .
+ *              but it could be any view model.
+ * Usage:
+ * <pre>
+ * fragment.getViewModelController(viewModelControllerFactory, ViewModel::class, arg)
+ * </pre>
+ *
+ * Note: When using proguard, add exclusion rules for your ViewModelControllerFactory
+ * <pre>
+ * -keepclassmembers class * implements com.acme.app.ViewModelControllerFactory { <methods>; }
+ * </pre>
+ */
+
+private const val KEY_SERIALIZED_DATA = "AndroidViewModelFactory.SerializedData"
+
+fun <T : ViewModel> Fragment.getViewModelController(
+    viewModelControllerFactory: Any,
+    requestedClass: KClass<T>
+): T {
+    return AndroidViewModelProviderFactory.with(
+        viewModelControllerFactory,
+        this,
+        extraSerializedData
+    )
+        .get(requestedClass.java)
+}
+
+fun <T : ViewModel> Fragment.getViewModelController(
+    viewModelControllerFactory: Any,
+    requestedClass: KClass<T>,
+    serializedData: String
+): T {
+    return AndroidViewModelProviderFactory.with(viewModelControllerFactory, this, serializedData)
+        .get(serializedData, requestedClass.java)
+}
+
+fun <T : ViewModel> FragmentActivity.getViewModelController(
+    viewModelControllerFactory: Any,
+    requestedClass: KClass<T>
+): T =
+    AndroidViewModelProviderFactory.with(viewModelControllerFactory, this, extraSerializedData)
+        .get(requestedClass.java)
+
+fun <T : ViewModel> FragmentActivity.getViewModelController(
```

Comment:
Why having the method on specifically `FragmentActivity`? Having it on `AppCompatActivity` seems more "standard" (even though AppCompatActivity extends FragmentActivity).

Reply:
Actually I'd rather go closer to the top of hierarchy than bottom of it. The absolute minimum I need is to be able to access the view model store. I think I'll go the other way around and use `ComponentActivity` rather than `FragmentActivity`.